Add ability to create druid responses

kevinhinterlong opened the pull request:
Needed to create druid responses from sql results in #364

kevinhinterlong committed:
Add ability to create druid responses

kevinhinterlong committed:
Explicitly create DateTime during test so timezone doesn't affect the results

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResponse.java:
```diff
@@ -0,0 +1,34 @@
+// Copyright 2016 Yahoo Inc.
```
2017

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResponse.java:
```diff
@@ -0,0 +1,34 @@
+ */
+@JsonSerialize(using = DruidResponseSerializer.class)
+public class DruidResponse {
+    private final List<DruidResultRow> results = new ArrayList<>();
```
It makes more sense to have a constructor that does the assignment.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResponseSerializer.java:
```diff
@@ -0,0 +1,26 @@
+// Copyright 2016 Yahoo Inc.
```
2017

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResponseSerializer.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.IOException;
+
+/**
+ * Serializes {@link DruidResponse} into a json equivalent to Druid's actual responses.
```
No sure what you mean by "Serializes X into Y to Z"

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResponseSerializer.java:
```diff
@@ -0,0 +1,26 @@
+public class DruidResponseSerializer extends JsonSerializer {
+
+    @Override
+    public void serialize(Object o, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
```
`o` doesn't tell what this variable does

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResponseSerializer.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public void serialize(Object o, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+            throws IOException {
```
I think JSON parsing exception should be handled inside the method, which means we should throw in a `try-catch`; you don't want to pass this method-logic to API callers.

kevinhinterlong replied:
Yea I was just following the same way as `DimensionToNameSerializer` did it.

As far as I can tell this logic never gets passed to callers directly (unless they instantiate a `DruidResponseSerializer`) because they go through the `ObjectMapper` which I'm guessing is handling the exception. 

Let me know if you still want me to catch it because I'm not sure which way is expected with jackson.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResultRow.java:
```diff
@@ -0,0 +1,39 @@
+// Copyright 2016 Yahoo Inc.
```
2017

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResultRow.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public abstract class DruidResultRow {
+    @JsonIgnore
+    public final DateTime timestamp;
```
Why do we make it public?

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResultRow.java:
```diff
@@ -0,0 +1,39 @@
+
+    @JsonProperty
+    public String getTimestamp() {
+        return timestamp.toDateTime(DateTimeZone.UTC).toString();
```
It's more efficient to move `timestamp.toDateTime(DateTimeZone.UTC).toString()` to a instance variable if this method will be called multiple times.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/GroupByResultRow.java:
```diff
@@ -0,0 +1,75 @@
+// Copyright 2016 Yahoo Inc.
```
2017

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/GroupByResultRow.java:
```diff
@@ -0,0 +1,75 @@
+@JsonPropertyOrder({"version", "timestamp", "event"})
+public class GroupByResultRow extends DruidResultRow {
+    @JsonIgnore
+    public final Version version;
```
Why do we make it public?

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/GroupByResultRow.java:
```diff
@@ -0,0 +1,75 @@
+    public final Version version;
+
+    @JsonProperty
+    private final Map<String, Object> event = new HashMap<>();
```
It makes more sense to have a constructor that does the assignment.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/TimeseriesResultRow.java:
```diff
@@ -0,0 +1,34 @@
+// Copyright 2016 Yahoo Inc.
```
2017

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/TimeseriesResultRow.java:
```diff
@@ -0,0 +1,34 @@
+@JsonPropertyOrder({"timestamp", "result"})
+public class TimeseriesResultRow extends DruidResultRow {
+    @JsonProperty
+    private final Map<String, Object> result = new HashMap<>();
```
Constructor

QubitPi commented on fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/response/DruidResponseSerializationSpec.groovy:
```diff
@@ -0,0 +1,61 @@
+package com.yahoo.bard.webservice.druid.model.response
```
```
// Copyright 2017 Yahoo Inc.
// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
```

QubitPi commented on fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/response/DruidResponseSerializationSpec.groovy:
```diff
@@ -0,0 +1,61 @@
+    static key2 = "key2"
+
+    @Unroll
+    def "Check Timeseries druid response serialization"() {
```
If Unrolls it, it's common to add a description for each different test case. For example https://github.com/yahoo/fili/blob/master/fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/MetricUnionAvailabilitySpec.groovy#L90

QubitPi commented on fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/response/DruidResponseSerializationSpec.groovy:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Unroll
+    def "Check GroupBy druid response serialization"() {
```
Description

kevinhinterlong committed:
Add TopN serialization and address comments

kevinhinterlong committed:
Remove unused imports

kevinhinterlong commented:
My comment got collapsed after making changes so I'll put it down here @QubitPi 

> Yea I was just following the same way as DimensionToNameSerializer did it.
> 
> As far as I can tell this logic never gets passed to callers directly (unless they instantiate a DruidResponseSerializer) because they go through the ObjectMapper which I'm guessing is handling the exception.
> 
> Let me know if you still want me to catch it because I'm not sure which way is expected with jackson.

Same thing with #399, I'm not sure what the expected behavior is for a `JsonSerializer`

kevinhinterlong committed:
Add javadoc comment on constructor

garyluoex commented:
@QubitPi  Does it make sense to catch it? If we catch it, what should we do in the catch block?

kevinhinterlong committed:
Add test case for empty string

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResultRow.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public abstract class DruidResultRow {
+    @JsonIgnore
+    private final String timestamp;
```
Having searched Fili codebase, I see timestamp is modeled using `DateTime`, instead of String, everywhere such as `Result`, `TopNResultSetMapper`, `DataSourceMetadataLoader`.

Having it as `DateTime` is also better so that child class can select different timezone instead of fixing to `DateTimeZone.UTC`

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/DruidResultRow.java:
```diff
@@ -0,0 +1,39 @@
+     * @param key  The key to be added.
+     * @param value  The value of the key.
+     */
+    public abstract void add(String key, Object value);
```
`public abstract void add(String key, V value);`

`public abstract class DruidResultRow<V> {`

kevinhinterlong replied:
This has to be able to store `String` and `Number` types. If you would like I could expand it to
```java
public abstract void add(String key, String value);
public abstract void add(String key, Number value);
```
but there is not a specific Type which can be associated with a row.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/GroupByResultRow.java:
```diff
@@ -0,0 +1,75 @@
+        V1("v1"),
+        V2("v2");
+
+        private String name;
```
`final`

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/response/GroupByResultRow.java:
```diff
@@ -0,0 +1,75 @@
+         *
+         * @param name  The name to be shown in results.
+         */
+        Version(String name) {
```
`final`

kevinhinterlong replied:
Did you mean this to be applied to and not the parameter?
```java
private String name;
```

kevinhinterlong committed:
Address comments

kevinhinterlong committed:
Add more tests and fix serialization of TopNResults

- TopN results are now combined by timestamps
- There is now a factory builder for DruidResponse to force using correct type

kevinhinterlong commented:
Druid responses are parsed into a result set and by wrapping all queries in a `AbstractDruidAggregationQuery` which overrides the query type to be `GROUP_BY` we can use a single format for druid responses (at least for the DruidResponseParser) so this is no longer needed.
